crow/recipes/binned_grid.py

```diff
@@ -0,0 +1,339 @@
+            self._completeness_distribution = self.completeness.distribution
+
+    def _setup_with_purity(self):
+        """Additional setup of class with the purity"""
+        if self.purity is None:
+            self._purity_distribution = self._flat_distribution
+        else:
+            self._purity_distribution = self.purity.distribution
+
+    def setup(self) -> None:
+        """Resets all internal dictionaries used for caching computed grids."""
+        self._hmf_grid = {}
+        self._mass_richness_grid = {}
+        self._completeness_grid = {}
+        self._purity_grid = {}
+        self._shear_grids = {}
+
+    def _get_hmf_grid(self, z: npt.NDArray[np.float64], sky_area: float, key):
+        """Compute HMF × comoving volume and store in the class."""
+
+        if key not in self._hmf_grid:
+            # sizes
+            n_m = len(self.log_mass_grid)
+            n_z = len(z)
+            # quantities
+            hmf_flat = self.cluster_theory.mass_function(
+                # flatten arrays to vectorize function
+                np.tile(self.log_mass_grid, n_z),
+                np.repeat(z, n_m),
+            )
+            mass_function_2d = hmf_flat.reshape(n_z, n_m)
+            vol = self.cluster_theory.comoving_volume(z, sky_area)
+            # assign
+            self._hmf_grid[key] = vol[:, np.newaxis] * mass_function_2d
+
+        return self._hmf_grid[key]
+
+    def _get_mass_richness_grid(
+        self, z: npt.NDArray[np.float64], log_proxy: npt.NDArray[np.float64], key
+    ):
+        """Compute mass-richness grid by vectorizing 1D inputs."""
+
+        if key not in self._mass_richness_grid:
+            # sizes
+            n_z = len(z)
+            n_m = len(self.log_mass_grid)
+            n_p = len(log_proxy)
+            # quantities
+            grid_3d_flat = self.mass_distribution.distribution(
+                # flatten arrays to vectorize function
+                np.tile(np.repeat(self.log_mass_grid, n_z), n_p),
+                np.tile(z, n_m * n_p),
+                np.repeat(log_proxy, n_z * n_m),
+            )
+            grid_3d_temp = grid_3d_flat.reshape(n_p, n_m, n_z)
+            # assign
+            self._mass_richness_grid[key] = grid_3d_temp.transpose(0, 2, 1)
+
+        return self._mass_richness_grid[key]
+
+    def _get_completeness_grid(self, z: npt.NDArray[np.float64], key):
+        """Compute completeness grid and store in the class."""
+
+        if key not in self._completeness_grid:
+            self._completeness_grid[key] = self._completeness_distribution(
+                self.log_mass_grid[np.newaxis, :], z[:, np.newaxis]
+            )
+
+        return self._completeness_grid[key]
+
+    def _get_purity_grid(
+        self, z: npt.NDArray[np.float64], log_proxy: npt.NDArray[np.float64], key
+    ):
+        """Compute purity grid and store in the class."""
+
+        if key not in self._purity_grid:
+            purity_yx = self._purity_distribution(
```

Comment:
we should pass `log_proxy[:, np.newaxis], z[np.newaxis, :]` instead of transposing the purity

Reply:
For this to work I had to change in line 74 of purity:
`if all(log_mass_proxy == -1.0):` to `if np.all(log_mass_proxy == -1.0):` because we were checking a first array with a scalar for `log_mass_proxy`  when the first input is `np.newaxis`